src/ModelContextProtocol.AspNetCore/Authentication/McpAuthenticationHandler.cs

```diff
@@ -0,0 +1,159 @@
+
+    /// <summary>
+    /// Gets the absolute URI for the resource metadata endpoint.
+    /// </summary>
+    private string GetAbsoluteResourceMetadataUri()
+    {
+        var options = _optionsMonitor.CurrentValue;
+        var resourceMetadataUri = options.ResourceMetadataUri;
+        
+        if (resourceMetadataUri.IsAbsoluteUri)
+        {
+            return resourceMetadataUri.ToString();
+        }
+        
+        // For relative URIs, combine with the base URL
+        string baseUrl = GetBaseUrl();
+        string resourceMetadataPath = resourceMetadataUri.ToString();
+        
+        if (!Uri.TryCreate(baseUrl + resourceMetadataPath, UriKind.Absolute, out var absoluteUri))
+        {
+            throw new InvalidOperationException("Could not create absolute URI for resource metadata.");
+        }
+        
+        return absoluteUri.ToString();
+    }
+
+    /// <summary>
+    /// Handles the resource metadata request.
+    /// </summary>
+    private async Task HandleResourceMetadataRequestAsync()
+    {
+        // Get resource metadata from options, using the dynamic provider if available
+        var options = _optionsMonitor.CurrentValue;
+        var resourceMetadata = options.GetResourceMetadata(Request.HttpContext);
+        
+        // Create a copy to avoid modifying the original
+        var metadata = new ProtectedResourceMetadata
+        {
+            AuthorizationServers = [.. resourceMetadata.AuthorizationServers],
+            BearerMethodsSupported = [.. resourceMetadata.BearerMethodsSupported],
+            ScopesSupported = [.. resourceMetadata.ScopesSupported],
+            ResourceDocumentation = resourceMetadata.ResourceDocumentation,
+            Resource = resourceMetadata.Resource
+        };
+        
+        // Set default resource if not set
+        if (metadata.Resource == null)
+        {
+            metadata.Resource = new Uri(GetBaseUrl());
+        }
+        
+        Response.StatusCode = StatusCodes.Status200OK;
+        Response.ContentType = "application/json";
+        
+        var json = System.Text.Json.JsonSerializer.Serialize(
+            metadata, 
+            McpJsonUtilities.DefaultOptions.GetTypeInfo(typeof(ProtectedResourceMetadata)));
+        
+        await Response.WriteAsync(json);
+    }
+
+    /// <inheritdoc />
+    protected override Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        // This handler doesn't perform authentication - it only adds resource metadata to challenges
+        // The actual authentication will be handled by the bearer token handler or other authentication handlers
+        return Task.FromResult(AuthenticateResult.NoResult());
+    }
+
+    /// <inheritdoc />
+    protected override Task HandleChallengeAsync(AuthenticationProperties properties)
+    {
+        // Set the response status code
+        Response.StatusCode = StatusCodes.Status401Unauthorized;
+
```

Comment:
```suggestion
```

Nit: base.HandleChallengeAsync already sets the 401 status code. It's okay if you prefer to be explicit. But in that case, I would call `base.HandleChallengeAsync(properties)` before this. Because as of now, this is just getting overwritten.